compiler/rustc_hir_typeck/src/_match.rs

```diff
@@ -77,12 +77,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let mut prior_non_diverging_arms = vec![]; // Used only for diagnostics.
         let mut prior_arm = None;
         for arm in arms {
+            self.diverges.set(Diverges::Maybe);
+
             if let Some(e) = &arm.guard {
-                self.diverges.set(Diverges::Maybe);
                 self.check_expr_has_type_or_error(e, tcx.types.bool, |_| {});
+
+                // FIXME: If this is the first arm and the pattern is irrefutable,
```

Comment:
This would be the place to insert a fix for #134734 (it'll need some extra divergence state tracking variable or something), though it also requires some rudimentary heuristic to compute if a pattern is irrefutable in HIR, since that's currently the responsibility of THIR pattern analysis.